gms/gossiper.cc

```diff
@@ -931,42 +942,30 @@ future<> gossiper::replicate_live_endpoints_on_change() {
     //
     // Gossiper task runs only on CPU0:
     //
-    //    - If endpoint_state_map or _live_endpoints have changed - replicate
-    //      them across all other shards.
-    //    - Reschedule the gossiper only after execution on all nodes is done.
+    //    - replicate _live_endpoints and _unreachable_endpoints
+    //      across all other shards.
+    //    - use _shadow_{live,unreachable}_endpoints_version on each shard
+    //      to determine if it has the latest copy, and replicate the respective
+    //      member from shard 0, if the shard is outdated.
     //
-    bool live_endpoint_changed = (_live_endpoints != _shadow_live_endpoints);
-    bool unreachable_endpoint_changed = (_unreachable_endpoints != _shadow_unreachable_endpoints);
-
-    if (live_endpoint_changed || unreachable_endpoint_changed) {
-        logger.debug("replicating live endpoints to other shards");
-
-        if (live_endpoint_changed) {
-            _shadow_live_endpoints = _live_endpoints;
-        }
-
-        if (unreachable_endpoint_changed) {
-            _shadow_unreachable_endpoints = _unreachable_endpoints;
-        }
-
-        co_await container().invoke_on_all([this, live_endpoint_changed, unreachable_endpoint_changed,
-                                   es = _endpoint_state_map] (gossiper& local_gossiper) {
-            // Don't copy gossiper(CPU0) maps into themselves!
-            if (this_shard_id() != 0) {
-                if (live_endpoint_changed) {
-                    local_gossiper._live_endpoints = _shadow_live_endpoints;
-                }
-
-                if (unreachable_endpoint_changed) {
-                    local_gossiper._unreachable_endpoints = _shadow_unreachable_endpoints;
-                }
+    logger.debug("replicating live endpoints to other shards");
+
+    co_await container().invoke_on_all([this] (gossiper& local_gossiper) {
+        // Don't copy gossiper(CPU0) maps into themselves!
+        if (this_shard_id() != 0) {
+            if (local_gossiper._shadow_live_endpoints_version != _live_endpoints_version) {
+                local_gossiper._live_endpoints = _live_endpoints;
+                local_gossiper._live_endpoints_version = _live_endpoints_version;
+                local_gossiper._shadow_live_endpoints_version = _live_endpoints_version;
+            }
 
-                for (auto&& e : es) {
-                    local_gossiper._endpoint_state_map[e.first].set_alive(e.second.is_alive());
-                }
+            if (local_gossiper._shadow_unreachable_endpoints_version != _unreachable_endpoints_version) {
+                local_gossiper._unreachable_endpoints = _unreachable_endpoints;
+                local_gossiper._unreachable_endpoints_version = _unreachable_endpoints_version;
+                local_gossiper._shadow_unreachable_endpoints_version = _unreachable_endpoints_version;
```

Comment:
I think the above replicating logic can be simplified and we can remove the whole shadow thing.
```
if (local_gossiper._live_endpoints_version = _live_endpoints_version) {
      local_gossiper._live_endpoints  =  _live_endpoints
      local_gossiper._unreachable_endpoints =  _unreachable_endpoints 
      local_gossiper._live_endpoints_version =  _live_endpoints_version
}
```

We can check if the non zero shard needs updating by checking if the non zero shard has the same version.
The replicating runs under the lock, so there is no danger that  the _live_endpoints or _unreahable_endpoints would change while shard zero is replicating to different non zero shards. 

Btw, like I mentioned in the other thread, we can combine live and shadow version.

Reply:
Since in this branch replicate_live_endpoints_on_change is called only under mutate_live_and_unreachable_endpoints it can simply bump the _live_endpoints_version and replicate _live_endpoints, _unreachable_endpoints, and _live_endpoints_version unconditionally.

Reply:
Yes! After the removal of the replicate_live_endpoints_on_change from gossiper::run.

Reply:
I was over-optimistic. I forgot it is also called from get_live_members_synchronized, and in this case we don't want to replicate unconditionally, so next version will just use _live_endpoints_version to detect a change, and it is bumped unconditionally in mutate_live_and_unreachable_endpoints to force an update on all other shards.